PyTorch Lightning DataParallel + ModelParallel benchmarks

lebrice opened the pull request:
Quick notes:
- Can easily switch precisions (16, 32, 64) 
- Can easily switch strategies (currently only doing dp and fsdp)
- Can switch between different profilers (simple, pytorch, etc):

breuleux commented on benchmarks/pytorch-lightning/README.md:
```diff
@@ -0,0 +1,4 @@
+
+# Benchmark
+
+Rewrite this README to explain what the benchmark is!
```
Doesn't have to be much, but describing the overall benchmark in a few lines would be handy.

lebrice replied:
Fixed in 47085da

breuleux commented on benchmarks/pytorch-lightning/giving_callback.py:
```diff
@@ -0,0 +1,95 @@
+""" PyTorch-Lightning callback that uses giving.give. Can be used on basically any model. """
```
Ohhh... that's neat. I might pull this into milabench proper somewhere so that the instrumentation could be reused for other lightning code.

breuleux commented on benchmarks/pytorch-lightning/giving_callback.py:
```diff
@@ -0,0 +1,95 @@
+    def on_train_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
+        self._end_time = datetime.datetime.now()
+        delta = self._end_time - self._start_time
+        give(length=delta)
```
Maybe `walltime`?

lebrice replied:
Fixed in 10c0bf2

breuleux commented on benchmarks/pytorch-lightning/utils.py:
```diff
@@ -0,0 +1,134 @@
+ModuleType = TypeVar("ModuleType", bound=nn.Module)
+
+
+# TODO maybe use this so we don't have to download the datasets, or use milatools/milavision.
+torchvision_dir: Path | None = Path("/network/datasets/torchvision")
+if not torchvision_dir.exists():
+    torchvision_dir = None
+
+DEFAULT_DATA_DIR: Path = Path(
+    os.environ.get("SLURM_TMPDIR", os.environ.get("SCRATCH", "data"))
+)
```
Not sure if this is used, but this is supposed to run anywhere (but typically from a Docker image), so no references to `/network`, `$SCRATCH` or `$SLURM_TMPDIR` can be made. The data has to live in a subdirectory of `$MILABENCH_DIR_DATA` and it should be downloaded there by the `prepare.py` script.

lebrice replied:
Fixed in 78f6867

breuleux commented on benchmarks/pytorch-lightning/main.py:
```diff
@@ -0,0 +1,172 @@
+        callbacks.append(GivingCallback())
+
+    print(f"Trainer kwargs:")
+    import pprint
```
Hoist imports to top, unless exceptional circumstances.

lebrice replied:
Fixed in 7f3383a

breuleux reviewed (commented):
I had no issues running `milabench install/prepare/run` on the cluster and I see the train_rate, so that's good.

Is there a way to make the output less verbose? The script appears to output percentage progress and stuff like that by default, but that mostly just clogs the logs.

I think the important remaining step is putting the data where milabench wants it to be (see comment).

lebrice commented:
> Is there a way to make the output less verbose? The script appears to output percentage progress and stuff like that by default, but that mostly just clogs the logs.

Fixed in 78f6867 
 
> I think the important remaining step is putting the data where milabench wants it to be (see comment).

Fixed in 4362316

breuleux commented on benchmarks/pytorch-lightning/main.py:
```diff
@@ -0,0 +1,171 @@
+#         limit_train_batches=100,
+#         limit_val_batches=100,
+#         profiler="simple",
+#     )
```
Clean up commented out code.

lebrice replied:
Fixed in 943a4da

breuleux commented on benchmarks/pytorch-lightning/prepare.py:
```diff
@@ -0,0 +1,31 @@
+    config = json.loads(os.environ["MILABENCH_CONFIG"])
+    data_directory = os.environ["MILABENCH_DIR_DATA"]
+    argv_dict = config["argv"]
+    argv_str = " ".join(f"--{k} {v}" for k, v in argv_dict.items())
```
I don't think that's going to work, `k` should already start with `--`. Also an entry such as `{"--flag": True}` should produce the string `--flag` and not `--flag True`, and if `v` is None it should be omitted.

You can import `milabench.multi._assemble_options` to make a list of option strings (and then you won't need shlex.split). I might move that to `milabench.utils`, but it'll have to do for now.

lebrice replied:
Fixed in d8498f3

breuleux commented on benchmarks/pytorch-lightning/utils.py:
```diff
@@ -0,0 +1,126 @@
+}
+
+
+# NOTE: This isn't
```
Don't leave us hanging. What is this not?? :)

lebrice replied:
Fixed in 0cef9ea

breuleux commented on benchmarks/pytorch-lightning/voirfile.py:
```diff
@@ -0,0 +1,35 @@
+# Import this to instrument the ArgumentParser, remove if no such thing
+from voir.overseer import Overseer
+
+from milabench.opt import instrument_argparse
```
Since you are using giving instead of the probes, keep only this line and remove the rest.

lebrice replied:
Fixed in e85a2f9b278000cae304ff2437c195a169159261
